src/crimson/osd/osd.cc

```diff
@@ -162,102 +162,116 @@ seastar::future<> OSD::open_meta_coll()
-	return seastar::now();
+      return store.create_new_collection(
+	coll_t::meta()
+      ).then([&store](auto ch) {
+	return OSDMeta(ch, store);
       });
     } else {
-      pg_shard_manager.init_meta_coll(ch, store);
-      return seastar::now();
+      return seastar::make_ready_future<OSDMeta>(ch, store);
     }
   });
 }
 
 seastar::future<> OSD::mkfs(
+  FuturizedStore &store,
+  unsigned whoami,
   uuid_d osd_uuid,
   uuid_d cluster_fsid,
   std::string osdspec_affinity)
 {
-  return store.start().then([this, osd_uuid] {
+  return store.start().then([&store, osd_uuid] {
     return store.mkfs(osd_uuid).handle_error(
       crimson::stateful_ec::handle([] (const auto& ec) {
         logger().error("error creating empty object store in {}: ({}) {}",
                        local_conf().get_val<std::string>("osd_data"),
                        ec.value(), ec.message());
         std::exit(EXIT_FAILURE);
       }));
-  }).then([this] {
+  }).then([&store] {
     return store.mount().handle_error(
-      crimson::stateful_ec::handle([] (const auto& ec) {
+      crimson::stateful_ec::handle([](const auto& ec) {
         logger().error("error mounting object store in {}: ({}) {}",
                        local_conf().get_val<std::string>("osd_data"),
                        ec.value(), ec.message());
         std::exit(EXIT_FAILURE);
       }));
-  }).then([this] {
-    return open_or_create_meta_coll();
-  }).then([cluster_fsid, this] {
+  }).then([&store] {
+    return open_or_create_meta_coll(store);
+  }).then([&store, whoami, cluster_fsid](auto meta_coll) {
+    OSDSuperblock superblock;
     superblock.cluster_fsid = cluster_fsid;
     superblock.osd_fsid = store.get_fsid();
     superblock.whoami = whoami;
     superblock.compat_features = get_osd_initial_compat_set();
-    return _write_superblock();
-  }).then([cluster_fsid, this] {
+    return _write_superblock(
+      store, std::move(meta_coll), std::move(superblock));
+  }).then([&store, cluster_fsid] {
     return store.write_meta("ceph_fsid", cluster_fsid.to_string());
-  }).then([this] {
+  }).then([&store] {
     return store.write_meta("magic", CEPH_OSD_ONDISK_MAGIC);
-  }).then([this] {
+  }).then([&store, whoami] {
     return store.write_meta("whoami", std::to_string(whoami));
-  }).then([this] {
-    return _write_key_meta();
-  }).then([this, osdspec_affinity=std::move(osdspec_affinity)] {
+  }).then([&store] {
+    return _write_key_meta(store);
+  }).then([&store, osdspec_affinity=std::move(osdspec_affinity)] {
     return store.write_meta("osdspec_affinity", osdspec_affinity);
-  }).then([this] {
+  }).then([&store] {
     return store.write_meta("ready", "ready");
-  }).then([cluster_fsid, this] {
+  }).then([&store, whoami, cluster_fsid] {
     fmt::print("created object store {} for osd.{} fsid {}\n",
                local_conf().get_val<std::string>("osd_data"),
                whoami, cluster_fsid);
-    return seastar::now();
+    return store.umount();
```

Comment:
Hmm, I don't see we were umounting before. If not, then – well – good to get this finally...

Reply:
Yep